### PR TITLE
Python base image: create airbyte user

### PR DESCRIPTION
**Pull request by alafanechere**
We want to make our python connector image use a non root user.
This PR cuts a new base image version which:
* Creates an `airbyte` user
* Make it the owner of `/airbyte`
* Make it able to read the cache dir and the path to the nltk data

### CODE REVIEW

**Comment by alafanechere** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/36544?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#36544** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/36544?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#36537** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/36537?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @alafanechere and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: airbyte-ci/connectors/base_images/base_images/hacks.py [hunk 1]
```diff
@@ -22,7 +22,9 @@ def get_container_dockerfile(container) -> str:
     """
 
     lineage = [
-        field for field in list(container._ctx.selections) if isinstance(field, dagger.api.base.Field) and field.type_name == "Container"
+        field
+        for field in list(container._ctx.selections)
+        if isinstance(field, dagger.client._core.Field) and field.type_name == "Container"
     ]
```
Comment (alafanechere) [inline airbyte-ci/connectors/base_images/base_images/hacks.py:28]: This change is due to a change in the latest dagger version

File: airbyte-ci/connectors/base_images/base_images/bases.py [hunk 2]
```diff
@@ -19,6 +19,10 @@ class AirbyteConnectorBaseImage(ABC):
     Please do not declare any Dagger with_exec instruction in this class as in the abstract class context we have no guarantee about the underlying system used in the base image.
```
Comment (erohmensing) [inline airbyte-ci/connectors/base_images/base_images/bases.py:19]: Does this concern still apply? I assume that the commands we're using are so basic as to not be concerned, but the messaging is confusing

File: airbyte-ci/connectors/base_images/base_images/python/bases.py [hunk 3]
```diff
@@ -67,7 +71,7 @@ def with_file_based_connector_dependencies(container: dagger.Container) -> dagge
             - nltk data
             """
             container = with_tesseract_and_poppler(container)
-            container = container.with_exec(["mkdir", self.nltk_data_path], skip_entrypoint=True).with_directory(
+            container = container.with_exec(["mkdir", "-p", "755", self.nltk_data_path], skip_entrypoint=True).with_directory(
```
Comment (erohmensing) [inline airbyte-ci/connectors/base_images/base_images/python/bases.py:74]: Optional: as we in the base class are doing the setup of the airbyte user, I think(?) that all new directories in the python base or other future bases have to be made with these permissions. Maybe in the base class we could define a method that we can use in the subsclasses to create based on the permissions required from the base.

File: airbyte-ci/connectors/base_images/base_images/sanity_checks.py [hunk 4]
```diff
@@ -99,3 +99,74 @@ async def check_socat_version(container: dagger.Container, expected_socat_versio
             raise errors.SanityCheckError(f"unexpected socat version: {version_number}")
     else:
         raise errors.SanityCheckError(f"Could not find the socat version in the version output: {socat_version_line}")
+
+
+async def check_user_exists(container: dagger.Container, user: str):
```
Comment (erohmensing) [inline airbyte-ci/connectors/base_images/base_images/sanity_checks.py:104]: Lots of sanity checks 🙏🏻

File: airbyte-ci/connectors/base_images/base_images/python/bases.py [hunk 5]
```diff
@@ -117,6 +118,12 @@ async def run_sanity_checks(self, platform: dagger.Platform):
         container = self.get_container(platform)
         await base_sanity_checks.check_timezone_is_utc(container)
         await base_sanity_checks.check_a_command_is_available_using_version_option(container, "bash")
+        await base_sanity_checks.check_user_exists(container, self.USER)
+        await base_sanity_checks.check_user_can_read_dir(container, self.USER, self.AIRBYTE_DIR_PATH)
+        await base_sanity_checks.check_user_can_read_dir(container, self.USER, self.nltk_data_path)
+        await base_sanity_checks.check_user_can_read_dir(container, self.USER, self.CACHE_DIR_PATH)
+        await base_sanity_checks.check_user_can_write_dir(container, self.USER, self.AIRBYTE_DIR_PATH)
+        await base_sanity_checks.check_user_cant_write_dir(container, self.USER, self.CACHE_DIR_PATH)
```
Comment (erohmensing) [inline airbyte-ci/connectors/base_images/base_images/python/bases.py:126]: Nit: For a second I thought 124 and 126 were contradicting. I think this is easier to parse when organized by file: 
```suggestion
        await base_sanity_checks.check_user_can_read_dir(container, self.USER, self.AIRBYTE_DIR_PATH)
        await base_sanity_checks.check_user_can_write_dir(container, self.USER, self.AIRBYTE_DIR_PATH)
        await base_sanity_checks.check_user_can_read_dir(container, self.USER, self.nltk_data_path)
        await base_sanity_checks.check_user_can_read_dir(container, self.USER, self.CACHE_DIR_PATH)
        await base_sanity_checks.check_user_cant_write_dir(container, self.USER, self.CACHE_DIR_PATH)
```

**Review by erohmensing — COMMENTED** [review]
I'll let someone else like Angel approve the functionality as I'm not comfortable saying I know all the pieces that need to change to make this change. Reviewing from a "update to base image side" and looks good

File: airbyte-ci/connectors/base_images/base_images/bases.py [hunk 6]
```diff
@@ -98,4 +102,16 @@ def get_base_container(self, platform: dagger.Platform) -> dagger.Container:
...
-        return self.dagger_client.pipeline(self.name_with_tag).container(platform=platform).from_(self.root_image.address)
+        return (
+            self.dagger_client.pipeline(self.name_with_tag)
+            .container(platform=platform)
+            .from_(self.root_image.address)
+            # Set the timezone to UTC
+            .with_exec(["ln", "-snf", "/usr/share/zoneinfo/Etc/UTC", "/etc/localtime"], skip_entrypoint=True)
+            # Install socat 1.7.4.4
+            .with_exec(["sh", "-c", "apt update && apt-get install -y socat=1.7.4.4-2"], skip_entrypoint=True)
+            .with_exec(["adduser", "--system", "--group", "--no-create-home", self.USER], skip_entrypoint=True)
```
Comment (perangel) [inline airbyte-ci/connectors/base_images/base_images/bases.py:113]:
```suggestion
            .with_exec(["adduser", "-u", "1000", "--system", "--group", "--no-create-home", self.USER], skip_entrypoint=True)
```

Comment (perangel) [reply airbyte-ci/connectors/base_images/base_images/bases.py:113]: We'd like to make the uid/gid consistent with the platform images (where `uid=1000(airbyte)` `gid=1000(airbyte)`)

**Comment by alafanechere** [comment]
### Merge activity

* **Nov 22, 3:42 AM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/airbytehq/airbyte/36544).

**Pull request merged**